doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,616 @@
+
+.. figure:: _static/nep-0040_dtype-hierarchy.png
+
+   **Figure:** Hierarchy of type objects reproduced from the reference
+   documentation. Some aliases such as ``np.intp`` are excluded. Datetime
+   and timedelta are not shown.
+
+NumPy scalars try to mimic zero dimensional arrays with a fixed datatype.
+For the numerical (and unicode) datatypes, they are further limited to
+native byte order.
+
+
+Current Implementation of Casting
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+One of the main features which datatypes need to support is casting between one
+another using ``arr.astype(new_dtype, casting="unsafe")``, or during execution
+of ufuncs with different types (such as adding integer and floating point numbers).
+
+Casting tables determine whether casting is possible or not.
+However, these cannot handle the parametric dtypes such as strings.
+The logic for parametric datatypes is defined mainly in ``PyArray_CanCastTo``
+and currently cannot be customized for user defined datatypes.
+
+The actual casting has two distinct parts:
+
+1. ``copyswap``/``copyswapn`` are defined for each dtype and can handle
+   byte-swapping for non-native byte orders as well as unaligned memory.
+2. The generic casting code is provided by C functions which know how to
+   cast aligned and contiguous memory from one dtype to another
+   (both in native byte order).
+3. C-level functions can be registered to cast aligned and contiguous memory
+   from one dtype to another.
+   The function may be provided with both arrays (although the parameter
+   is sometimes ``NULL`` for scalars).
+   NumPy will ensure that these functions receive native byte order input.
+   The current implementation stores the functions either in a C-array
+   on the datatype which is cast, or in a dictionary when casting to a user
+   defined datatype.
+
+When casting (small) buffers will be used when necessary to ensure
+contiguity, alignment or native byte order.
+In this case first ``copyswapn`` is called to and ensures that the cast function
+can handle the input.
+Generally NumPy will thus perform casting as chain of the three functions
+``in_copyswapn -> castfunc -> out_copyswapn``.
+
+The above multiple functions are wrapped into a single function (with metadata)
+that handles the cast and is used for example during the buffered iteration used
+by ufuncs.
+This is the mechanism that is always used for user defined datatypes.
+For most dtypes defined within NumPy itself, more specialized code is used to
+find define a function to do the actual cast
+(defined by the private ``PyArray_GetDTypeTransferFunction``).
+This mechanism replaces most of the above mechanism and provides much faster
+casts for for example when the inputs are not contiguous in memory.
+However, it cannot be extended by user defined datatypes.
+
+Related to casting, we currently have a ``PyArray_EquivTypes`` function which
+indicate that a *view* is sufficient (and thus no cast is necessary).
```

Comment:
Is this another example of functionality that is not available to user-defined types?